File: starter_kits/C++/hlt/player.cpp
#include "player.hpp"
#include "input.hpp"

void hlt::Player::_update(int num_ships, int num_dropoffs, Halite halite) {
    this->halite = halite;

    ships.clear();
    for (int i = 0; i < num_ships; ++i) {
        std::shared_ptr<hlt::Ship> ship = hlt::Ship::_generate(id);
        ships[ship->id] = ship;
    }

    dropoffs.clear();
    for (int i = 0; i < num_dropoffs; ++i) {
        std::shared_ptr<hlt::Dropoff> dropoff = hlt::Dropoff::_generate(id);
        dropoffs[dropoff->id] = dropoff;
    }
}

std::shared_ptr<hlt::Player> hlt::Player::_generate() {
    PlayerId player_id;
    int shipyard_x;
    int shipyard_y;
    hlt::get_sstream() >> player_id >> shipyard_x >> shipyard_y;

    return std::make_shared<hlt::Player>(player_id, shipyard_x, shipyard_y);
}

File: starter_kits/Rust/src/hlt/player.rs
use hlt::dropoff::Dropoff;
use hlt::DropoffId;
use hlt::input::Input;
use hlt::PlayerId;
use hlt::position::Position;
use hlt::ship::Ship;
use hlt::ShipId;
use hlt::shipyard::Shipyard;
use std::collections::HashMap;

pub struct Player {
    pub id: PlayerId,
    pub shipyard: Shipyard,
    pub halite: usize,
    pub ship_ids: Vec<ShipId>,
    pub dropoff_ids: Vec<DropoffId>,
}

impl Player {
    pub fn update(
        &mut self,
        input: &mut Input,
        max_halite: usize,
        ships: &mut HashMap<ShipId, Ship>,
        dropoffs: &mut HashMap<DropoffId, Dropoff>,
        num_ships: usize,
        num_dropoffs: usize,
        halite: usize)
    {
        self.halite = halite;

        self.ship_ids.clear();
        for _ in 0..num_ships {
            let ship = Ship::generate(input, self.id, max_halite);
            self.ship_ids.push(ship.id);
            ships.insert(ship.id, ship);
        }

        self.dropoff_ids.clear();
        for _ in 0..num_dropoffs {
            let dropoff = Dropoff::generate(input, self.id);
            self.dropoff_ids.push(dropoff.id);
            dropoffs.insert(dropoff.id, dropoff);
        }
    }

    pub fn generate(input: &mut Input) -> Player {
        input.read_and_parse_line();
        let id = PlayerId(input.next_usize());
        let shipyard_x = input.next_i32();
        let shipyard_y = input.next_i32();

        let shipyard = Shipyard { owner: id, position: Position { x: shipyard_x, y: shipyard_y } };

        Player { id, shipyard, halite: 0, ship_ids: Vec::new(), dropoff_ids: Vec::new() }
    }
}
